core/pubnub_revoke_token.c

```diff
@@ -0,0 +1,67 @@
+/* -*- c-file-style:"stroustrup"; indent-tabs-mode: nil -*- */
+#include "pubnub_internal.h"
+
+#include "core/pubnub_ccore.h"
+#include "core/pubnub_netcore.h"
+#include "core/pubnub_assert.h"
+#include "core/pubnub_timers.h"
+#include "core/pubnub_log.h"
+#include "lib/pb_strnlen_s.h"
+#include "pbcc_revoke_token.h"
+#include "pubnub_revoke_token.h"
+#include "core/pubnub_api_types.h"
+
+#include "core/pbpal.h"
+
+
+enum pubnub_res pubnub_revoke_token(pubnub_t* pb, char const* token)
+{
+    enum pubnub_res rslt;
+    
+    pubnub_mutex_lock(pb->monitor);
+    if(!pbnc_can_start_transaction(pb)) {
+        pubnub_mutex_unlock(pb->monitor);
+        return PNR_IN_PROGRESS;
+    }
+
+    pb->method = pubnubUseDELETE;
+    pb->trans = PBTT_REVOKE_TOKEN;
+
+    rslt = pbcc_revoke_token_prep(&pb->core, token, pb->trans);
```

Comment:
@650elx It doesn't look that `pbcc_revoke_token_prep` need information about HTTP method or transaction type. Let's reduce it a bit?

```c
pb->method = pubnubUseDELETE; // Remove this line
pb->trans = PBTT_REVOKE_TOKEN; // Remove this line

rslt = pbcc_revoke_token_prep(&pb->core, token, PBTT_REVOKE_TOKEN);
```

Those fields will be set in following condition, if parameteres preparation was successful.

Reply:
@parfeon You're right; I still have more things like this to optimise :)

Reply:
I would prefer to restore these lines. In `pubnub_get_revoke_token_response` method, we are using pb->trans.
```
pb->method = pubnubUseDELETE; // Remove this line
pb->trans = PBTT_REVOKE_TOKEN; // Remove this line
```

Reply:
@budgetpreneur those value will be set if `rslt = pbcc_revoke_token_prep(&pb->core, token, pb->trans);` will be successful.

Reply:
@parfeon what happens if not successful?

Reply:
@budgetpreneur if `pubnub_revoke_token` call fails there is no point in attempt to use `pubnub_get_revoke_token_response`. This behavior is the same for all API inside of C-core SDK.

Reply:
In the past, during my play with sdk, I got some confusion results. I don't exactly recall.  Once I called publish(for example) and here_now  and tried to call `pubnub_get`. Here when the first one is success and the second one failed, there could be potential chance of getting invalid results, if SDK don't check for which request, the response was expected.
So here in `pubnub_get_revoke_token_response`, SDK was comparing if valid transaction type, before sending response. This behaviour was not added before and it started with grant_token. fetch_history(PR ready).

Reply:
@budgetpreneur I was referring to default SDK usage example:
```c
res = pubnub_publish(pbp_2, chan, "\"Hello world from subscribe-publish callback sample!\"");
if (res != PNR_STARTED) {
  printf("pubnub_publish() returned unexpected: %d('%s')\n", res, pubnub_res_2_string(res));
  callback_sample_free(pbp);
  callback_sample_free(pbp_2);
  return -1;
}
```

Single PubNub context can handle only one request at a time. So there shouldn't be confusion about which request failed.

Reply:
I am aware that c-core supports single pubnub context. I thought checking the trans type while sending response would drive some developer experience, especially for new developers, when they code like this. 
    ```
enum pubnub_res hn_res = pubnub_here_now(pbp, chan, NULL);
    if (PNR_STARTED == hn_res) {
        hn_res = pubnub_await(pbp);
    }

    enum pubnub_res pub_res = pubnub_publish(pbp, chan, "\"Hello world from sync!\"");
    if (PNR_STARTED == pub_res) {
        pub_res = pubnub_await(pbp);
    }

    if (PNR_OK == pub_res) {
        printf("TEST Published! Response from Pubnub: %s\n",
               pubnub_last_publish_result(pbp));
    }

    if (PNR_OK == hn_res) {
        puts("TEST Got here now presence!");
        for (;;) {
            msg = pubnub_get(pbp);
            if (NULL == msg) {
                break;
            }
            puts(msg);
        }
    }

```

Reply:
@budgetpreneur example won't work as expected from the starts, even if suggested lines returned back, because:
```c
if (PNR_OK == hn_res) {
    puts("TEST Got here now presence!");
    for (;;) {
        msg = pubnub_get(pbp);
        if (NULL == msg) {
            break;
        }
        puts(msg);
    }
}
```

Will work with buffer, which contains results of `pubnub_publish(pbp, chan, "\"Hello world from sync!\"");`

Reply:
Its weird, I see this code, where pb->trans value not assigned, Can someone confirm?
```
enum pubnub_res pubnub_revoke_token(pubnub_t* pb, char const* token)
{
    enum pubnub_res rslt;
    
    pubnub_mutex_lock(pb->monitor);
    if(!pbnc_can_start_transaction(pb)) {
        pubnub_mutex_unlock(pb->monitor);
        return PNR_IN_PROGRESS;
    }

    rslt = pbcc_revoke_token_prep(&pb->core, token, pb->trans);

    if (PNR_STARTED == rslt) {
        pb->trans            = PBTT_REVOKE_TOKEN;
        pb->core.last_result = PNR_STARTED;
        pb->method           = pubnubUseDELETE;
        pbnc_fsm(pb);
        rslt = pb->core.last_result;
    }

    pubnub_mutex_unlock(pb->monitor);

    return rslt;
}
```

Reply:
@budgetpreneur looks like you pulled out latest code, where Bartosz changed it.
Do you still think, that it is wrong approach to do it all in the same way as other APIs does?

Reply:
If feature is working as expected, we can keep it. No issue.